### ISSUE TITLE
How are metrics aggregated in DDP?

### ISSUE THREAD

**Issue by ZhaofengWu** [PyTorchLightning/pytorch-lightning#2109]
I see that the `LightningTemplateModel` in `0.7.5` (no longer the case in `master`) manually averages the metrics in `validation_epoch_end` for DP and DDP2
https://github.com/PyTorchLightning/pytorch-lightning/blob/694f1d789dfa56b365b68dd4f3c6f5f7a4c8970a/pl_examples/models/lightning_template.py#L167-L168

But what about DDP? I get that each device can have its own loss for backward, but we want only one single metric across devices. How is that achieved? (Is averaging the best way to aggregate most metrics anyway?)

**Comment by ZhaofengWu** [comment]
My guess was that only the train dataloader uses `DistributedSampler` but not val/test. In other words each process evals the entire val/test sets and only rank 0 reports (e.g. logs) the metrics. Apparently this used to be the case but #1192 changed val/test sets to use `DistributedSampler` too. So I think some aggregation must be done?

**Comment by ZhaofengWu** [comment]
I found this
https://github.com/PyTorchLightning/pytorch-lightning/blob/bd49b07fbba09b1e7d8851ee5a1ffce3d5925e9e/pytorch_lightning/metrics/metric.py#L46-L54
But if I don't want the overhead of creating a class for simple one-liner metrics, and/or have metrics that can't be easily reduced, is there a way to let dev/test dataloaders to load the entire datasets like pre-#1192? The only way I can think of is to set `replace_sampler_ddp=False` and manually add the `DistributedSampler` to the training dataloader. That feels kind of hacky though. If there's an option like `replace_evaluation_sampler_ddp` it would be much more straightforward.

**Comment by alexeykarnachev** [comment]
@ZhaofengWu could you, please provide a min. runnable script, which represents the problem?

**Comment by ZhaofengWu** [comment]
Sorry but it's not a problem/bug in the code. It's just a question: what's the proper way to aggregate metrics under DDP if we don't want the overhead of subclassing the `TensorMetric` mentioned above. If "letting dev/test dataloaders read the entire datasets" is the answer, what's the best way to do that?

**Comment by aaronma2020** [comment]
I have the same problem

**Comment by Borda** [comment]
@aaronma2020 mind provide minimal running example?
